Stop taking locks when reading from unique indexes

systay opened the pull request.

Commit:
Stop taking locks when reading from unique indexes (Cost planner)

Commit:
Stop taking locks when reading from unique indexes (Rule planner)

Commit:
Renamed QueryContext method to signify what it actually does

systay commented:
When merging this into 3.0, we should use this:
https://github.com/pontusmelke/neo4j/commit/33d399c19a8d80d5f80a6564422b3ea2e3195b92

It makes sure the cost planner doesn't take locks when doing read-only access.

pontusmelke commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/UniqueIndexUsageAcceptanceTest.scala:
```diff
@@ -0,0 +1,144 @@
+    val result = executeWithAllPlanners("MATCH (n:Crew) WHERE n.name = 'Neo' RETURN n")
+
+    // Then
+    result.executionPlanDescription().toString should include("NodeUniqueIndexSeek")
```
~~You can use `result should use("NodeUniqueIndexSeek")`~~

~~(likewise for all tests in this class)~~

Probably not in 2.2

henriknyman committed:
Add test for no locking when using index search

Tests that no locking happens when using index search in read-only queries
(using a new monitor).

lutovich commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_2/TransactionBoundQueryContext.scala:
```diff
@@ -131,10 +132,13 @@ final class TransactionBoundQueryContext(graph: GraphDatabaseAPI,
+  }
 
-  def exactUniqueIndexSearch(index: IndexDescriptor, value: Any): Option[Node] = {
+  def lockingIndexSearch(index: IndexDescriptor, value: Any): Option[Node] = {
```
Naming of these methods is a bit inconsistent. Maybe call this one `lockingExactIndexSearch`?
Or maybe we can name them `exactIndexSearchForRead` and `exactIndexSearchForWrite`.

systay commented:
merged manually
